File: ios/SseConnection.m
#import "SseConnection.h"

// ─── SSE line-state machine ───────────────────────────────────────────────────

typedef NS_ENUM(NSInteger, SSELineState) {
  SSELineStateNormal,   // reading ordinary characters
  SSELineStateAfterCR,  // just saw a bare \r, waiting to see if \n follows
};

// ─── SseConnection implementation ────────────────────────────────────────────

@implementation SseConnection {
  NSURL *_url;
  NSString *_method;
  NSDictionary<NSString *, NSString *> *_headers;
  NSString *_body;
  NSTimeInterval _timeout;

  SseOpenBlock  _onOpen;
  SseEventBlock _onEvent;
  SseErrorBlock _onError;
  SseCloseBlock _onClose;

  NSURLSession *_session;
  NSURLSessionDataTask *_task;
  dispatch_queue_t _callbackQueue;

  // Protects _cancelled and connection lifecycle.
  dispatch_semaphore_t _lock;
  BOOL _cancelled;
  BOOL _openFired;

  // ─── Line buffer ────────────────────────────────────────────────────────────
  // Accumulates bytes for the current (possibly incomplete) line.
  NSMutableData *_lineData;
  SSELineState _lineState;

  // ─── SSE parser state ───────────────────────────────────────────────────────
  NSString *_eventType;
  NSMutableArray<NSString *> *_dataLines;
  NSString *_lastEventId;
  NSNumber *_retry;
}

@synthesize streamId = _streamId;
NSString *_streamId;

// ─── Initialisation ───────────────────────────────────────────────────────────

- (instancetype)initWithStreamId:(NSString *)streamId
                             url:(NSURL *)url
                          method:(NSString *)method
                         headers:(NSDictionary<NSString *, NSString *> *)headers
                            body:(nullable NSString *)body
                         timeout:(NSTimeInterval)timeout
                          onOpen:(SseOpenBlock)onOpen
                         onEvent:(SseEventBlock)onEvent
                         onError:(SseErrorBlock)onError
                         onClose:(SseCloseBlock)onClose
{
  if (self = [super init]) {
    _streamId = [streamId copy];
    _url      = url;
    _method   = [method copy];
    _headers  = [headers copy];
    _body     = [body copy];
    _timeout  = timeout;

    _onOpen  = [onOpen copy];
    _onEvent = [onEvent copy];
    _onError = [onError copy];
    _onClose = [onClose copy];

    _lock          = dispatch_semaphore_create(1);
    _callbackQueue = dispatch_queue_create("com.nativesse.connection", DISPATCH_QUEUE_SERIAL);

    _lineData   = [NSMutableData data];
    _lineState  = SSELineStateNormal;
    _dataLines  = [NSMutableArray array];
    _eventType  = @"";
    _lastEventId = @"";
    _retry      = nil;
  }
  return self;
}

// ─── Connect / Disconnect ─────────────────────────────────────────────────────

- (void)connect
{
  dispatch_semaphore_wait(_lock, DISPATCH_TIME_FOREVER);
  _cancelled = NO;
  _openFired = NO;
  dispatch_semaphore_signal(_lock);

  NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:_url];
  request.HTTPMethod = _method;

  for (NSString *key in _headers) {
    [request setValue:_headers[key] forHTTPHeaderField:key];
  }

  if (_body.length > 0 && ![_method isEqualToString:@"GET"]) {
    request.HTTPBody = [_body dataUsingEncoding:NSUTF8StringEncoding];
  }

  if (_timeout > 0) {
    request.timeoutInterval = _timeout / 1000.0;
  }

  // Disable caching – essential for SSE.
  request.cachePolicy = NSURLRequestReloadIgnoringLocalCacheData;

  NSURLSessionConfiguration *cfg = [NSURLSessionConfiguration defaultSessionConfiguration];
  cfg.requestCachePolicy          = NSURLRequestReloadIgnoringLocalCacheData;
  // SSE streams are long-lived; URLSession must not time out on idle.
  cfg.timeoutIntervalForRequest   = (_timeout > 0) ? _timeout / 1000.0 : DBL_MAX;
  cfg.timeoutIntervalForResource  = DBL_MAX;

  // Delegate queue is nil → URLSession uses its own background thread.
  _session = [NSURLSession sessionWithConfiguration:cfg delegate:self delegateQueue:nil];
  _task    = [_session dataTaskWithRequest:request];
  [_task resume];
}

- (void)disconnect
{
  dispatch_semaphore_wait(_lock, DISPATCH_TIME_FOREVER);
  _cancelled = YES;
  dispatch_semaphore_signal(_lock);

  [_task cancel];
  [_session invalidateAndCancel];
  _task    = nil;
  _session = nil;
}

// ─── NSURLSessionDataDelegate ─────────────────────────────────────────────────

- (void)URLSession:(NSURLSession *)session
          dataTask:(NSURLSessionDataTask *)dataTask
didReceiveResponse:(NSURLResponse *)response
 completionHandler:(void (^)(NSURLSessionResponseDisposition))completionHandler
{
  dispatch_semaphore_wait(_lock, DISPATCH_TIME_FOREVER);
  BOOL cancelled = _cancelled;
  dispatch_semaphore_signal(_lock);

  if (cancelled) {
    completionHandler(NSURLSessionResponseCancel);
    return;
  }

  if (![response isKindOfClass:[NSHTTPURLResponse class]]) {
    completionHandler(NSURLSessionResponseCancel);
    NSString *sid = _streamId;
    dispatch_async(_callbackQueue, ^{
      self->_onError(@"Non-HTTP response", nil, YES);
    });
    (void)sid;
    return;
  }

  NSHTTPURLResponse *http = (NSHTTPURLResponse *)response;
  NSInteger statusCode = http.statusCode;

  if (statusCode < 200 || statusCode >= 300) {
    completionHandler(NSURLSessionResponseCancel);
    NSNumber *code = @(statusCode);
    NSString *msg  = [NSString stringWithFormat:@"HTTP error %ld", (long)statusCode];
    dispatch_async(_callbackQueue, ^{
      self->_onError(msg, code, YES);
    });
    return;
  }

  // Collect response headers.
  NSMutableDictionary<NSString *, NSString *> *respHeaders = [NSMutableDictionary dictionary];
  [http.allHeaderFields enumerateKeysAndObjectsUsingBlock:^(id key, id val, BOOL *__unused stop) {
    respHeaders[[key description]] = [val description];
  }];

  _openFired = YES;
  NSDictionary<NSString *, NSString *> *frozen = [respHeaders copy];
  dispatch_async(_callbackQueue, ^{
    self->_onOpen(statusCode, frozen);
  });

  completionHandler(NSURLSessionResponseAllow);
}

- (void)URLSession:(NSURLSession *)session
          dataTask:(NSURLSessionDataTask *)dataTask
    didReceiveData:(NSData *)data
{
  dispatch_semaphore_wait(_lock, DISPATCH_TIME_FOREVER);
  BOOL cancelled = _cancelled;
  dispatch_semaphore_signal(_lock);
  if (cancelled) return;

  [self processBytes:(const uint8_t *)data.bytes length:data.length];
}

- (void)URLSession:(NSURLSession *)session
              task:(NSURLSessionTask *)task
didCompleteWithError:(nullable NSError *)error
{
  dispatch_semaphore_wait(_lock, DISPATCH_TIME_FOREVER);
  BOOL cancelled = _cancelled;
  dispatch_semaphore_signal(_lock);
  if (cancelled) return;

  if (error) {
    if (error.code == NSURLErrorCancelled) return;
    NSString *msg  = error.localizedDescription;
    NSNumber *code = @(error.code);
    dispatch_async(_callbackQueue, ^{
      self->_onError(msg, code, NO);
    });
  } else {
    dispatch_async(_callbackQueue, ^{
      self->_onClose();
    });
  }
}

// ─── Byte-stream SSE line parser ──────────────────────────────────────────────

/**
 * Process raw incoming bytes through the SSE line-state machine.
 *
 * SSE specifies that lines are terminated by:
 *   U+000A LF
 *   U+000D CR
 *   U+000D CR followed by U+000A LF  (treated as a single terminator)
 *
 * We handle the CR→LF boundary-across-chunks case with _lineState.
 */
- (void)processBytes:(const uint8_t *)bytes length:(NSUInteger)length
{
  for (NSUInteger i = 0; i < length; i++) {
    uint8_t c = bytes[i];

    switch (_lineState) {
      case SSELineStateNormal:
        if (c == '\n') {
          [self emitLine];
        } else if (c == '\r') {
          [self emitLine];
          _lineState = SSELineStateAfterCR;
        } else {
          [_lineData appendBytes:&c length:1];
        }
        break;

      case SSELineStateAfterCR:
        if (c == '\n') {
          // \r\n: the line was already emitted on \r; skip this \n.
        } else if (c == '\r') {
          // Another bare \r immediately after: emit an empty line and stay.
          [self emitLine];
          // Stay in SSELineStateAfterCR.
        } else {
          // Not a \n after \r → start new line with this character.
          [_lineData appendBytes:&c length:1];
        }
        _lineState = (c == '\r') ? SSELineStateAfterCR : SSELineStateNormal;
        break;
    }
  }
}

/** Called with the complete bytes for one SSE line (without the terminator). */
- (void)emitLine
{
  NSString *line = [[NSString alloc] initWithData:_lineData encoding:NSUTF8StringEncoding];
  if (!line) {
    // Gracefully handle invalid UTF-8: treat as empty line.
    line = @"";
  }
  [_lineData setLength:0];
  [self parseLine:line];
}

// ─── SSE field parser ─────────────────────────────────────────────────────────

- (void)parseLine:(NSString *)line
{
  if (line.length == 0) {
    [self dispatchCurrentEvent];
    return;
  }

  if ([line hasPrefix:@":"]) {
    return; // Comment – ignored.
  }

  NSString *field;
  NSString *value;
  NSRange colonRange = [line rangeOfString:@":"];

  if (colonRange.location == NSNotFound) {
    field = line;
    value = @"";
  } else {
    field = [line substringToIndex:colonRange.location];
    NSString *raw = [line substringFromIndex:colonRange.location + 1];
    // Strip a single leading space per spec.
    value = [raw hasPrefix:@" "] ? [raw substringFromIndex:1] : raw;
  }

  if ([field isEqualToString:@"event"]) {
    _eventType = value;
  } else if ([field isEqualToString:@"data"]) {
    [_dataLines addObject:value];
  } else if ([field isEqualToString:@"id"]) {
    // Per spec: ignore if value contains U+0000 NULL.
    if ([value rangeOfString:@"\0"].location == NSNotFound) {
      _lastEventId = value;
    }
  } else if ([field isEqualToString:@"retry"]) {
    // Must be all ASCII digits.
    NSCharacterSet *nonDigit = [[NSCharacterSet decimalDigitCharacterSet] invertedSet];
    if ([value rangeOfCharacterFromSet:nonDigit].location == NSNotFound && value.length > 0) {
      NSInteger ms = [value integerValue];
      _retry = @(ms);
      NSNumber *retryNum = _retry;
      dispatch_async(_callbackQueue, ^{
        self->_onEvent(@"__retry__", @"", @"", retryNum);
      });
    }
  }
  // Unknown fields are silently ignored.
}

- (void)dispatchCurrentEvent
{
  if (_dataLines.count == 0) {
    // Empty data buffer → reset event fields and do not dispatch.
    _eventType = @"";
    return;
  }

  NSString *data        = [_dataLines componentsJoinedByString:@"\n"];
  NSString *type        = (_eventType.length > 0) ? _eventType : @"message";
  NSString *lastEventId = [_lastEventId copy];
  NSNumber *retry       = _retry;

  // Reset event-scoped fields (lastEventId persists globally per spec).
  _eventType = @"";
  _dataLines = [NSMutableArray array];
  _retry     = nil;

  dispatch_async(_callbackQueue, ^{
    self->_onEvent(type, data, lastEventId, retry);
  });
}

@end

File: ios/SseConnection.swift
// SseConnection.swift
// Swift implementation using URLSession streaming data tasks.
//
// Design:
//  • Thin transport: raw UTF-8 chunks are forwarded to JS for SSE parsing.
//    All SSE protocol parsing (line splitting, field extraction, event dispatch)
//    lives in the JS SseParser, keeping native code minimal.
//  • Decoding is incremental: URLSession hands over arbitrary byte boundaries,
//    so a partial multi-byte sequence is held back until its remaining bytes
//    arrive. Decoding each delivery in isolation would corrupt — or, with
//    `String(data:encoding:)`, silently discard — any non-ASCII stream.
//  • Exposed to ObjC++ via @objcMembers so NativeSse.mm can use it directly.

import Foundation

// MARK: - SseConnectionSwift

@objcMembers
public final class SseConnectionSwift: NSObject {

    // MARK: Public properties

    public let streamId: String

    // Callbacks set by NativeSse.mm before connect() is called.
    public var onOpen:  ((Int, [String: String]) -> Void)?
    /// Parameters: (decodedText, byteLength)
    public var onChunk: ((String, Int) -> Void)?
    /// Parameters: (message, statusCode or -1, errorCode, isFatal)
    public var onError: ((String, Int, String, Bool) -> Void)?
    public var onClose: (() -> Void)?

    // MARK: Private configuration

    private let url: URL
    private let method: String
    private let headers: [String: String]
    private let body: String?
    private let timeoutMs: TimeInterval   // in ms; 0 = no timeout

    // MARK: Private networking state

    private var session: URLSession?
    private var task: URLSessionDataTask?

    // NSLock rather than os_unfair_lock: taking `&_lock` on a stored property of
    // a class does not guarantee a stable address in Swift, so the lock may not
    // actually lock. This is taken once per delivery — the cost is noise next to
    // the bridge crossing that follows.
    private let lock = NSLock()
    private var _cancelled = false
    private var _didOpen = false
    private var _connectTimer: DispatchWorkItem?

    /// Bytes received but not yet decodable: the tail of a split UTF-8 sequence.
    /// Never exceeds 3 bytes. Only touched on the delegate queue.
    private var pending = Data()

    // MARK: Init

    public init(
        streamId:  String,
        url:       URL,
        method:    String,
        headers:   [String: String],
        body:      String?,
        timeoutMs: Double
    ) {
        self.streamId  = streamId
        self.url       = url
        self.method    = method
        self.headers   = headers
        self.body      = body
        self.timeoutMs = timeoutMs
        super.init()
    }

    // MARK: Public API

    public func connect() {
        lock.lock()
        _cancelled = false
        _didOpen = false
        pending.removeAll(keepingCapacity: true)
        lock.unlock()

        var request = URLRequest(url: url)
        request.httpMethod = method
        headers.forEach { request.setValue($1, forHTTPHeaderField: $0) }

        if let b = body, !b.isEmpty, method.uppercased() != "GET" {
            request.httpBody = b.data(using: .utf8)
        }
        request.cachePolicy = .reloadIgnoringLocalCacheData

        let cfg = URLSessionConfiguration.default
        cfg.requestCachePolicy = .reloadIgnoringLocalCacheData
        // `timeoutIntervalForRequest` is an inactivity timeout, and silence is
        // normal on an SSE stream. Zombie detection belongs to the JS
        // `staleTimeoutMs` option, which knows whether events are arriving.
        cfg.timeoutIntervalForRequest  = .greatestFiniteMagnitude
        cfg.timeoutIntervalForResource = .greatestFiniteMagnitude

        // delegateQueue: nil → URLSession uses its own background serial queue.
        session = URLSession(configuration: cfg, delegate: self, delegateQueue: nil)
        task = session!.dataTask(with: request)
        if timeoutMs > 0 {
            // Approximate a connect timeout: URLSession has no dedicated one, so
            // bound how long the task may sit before a response arrives.
            scheduleConnectTimeout(seconds: timeoutMs / 1_000)
        }
        task!.resume()
    }

    public func disconnect() {
        lock.lock()
        _cancelled = true
        let timer = _connectTimer
        _connectTimer = nil
        let s = session
        let t = task
        session = nil
        task = nil
        lock.unlock()

        timer?.cancel()
        t?.cancel()
        // Releases the session's strong reference to this delegate.
        s?.invalidateAndCancel()
    }

    // MARK: Private helpers

    private var isCancelled: Bool {
        lock.lock()
        defer { lock.unlock() }
        return _cancelled
    }

    private func scheduleConnectTimeout(seconds: TimeInterval) {
        let item = DispatchWorkItem { [weak self] in
            guard let self else { return }
            self.lock.lock()
            let fire = !self._cancelled && !self._didOpen
            self.lock.unlock()
            guard fire else { return }
            self.onError?("Connection timed out", -1, "TIMEOUT_ERROR", false)
            self.teardown()
        }
        lock.lock(); _connectTimer = item; lock.unlock()
        DispatchQueue.global().asyncAfter(deadline: .now() + seconds, execute: item)
    }

    /// Mark the response as arrived and cancel the connect deadline.
    private func markOpened() {
        lock.lock()
        _didOpen = true
        let timer = _connectTimer
        _connectTimer = nil
        lock.unlock()
        timer?.cancel()
    }

    /// Release the session so it stops retaining this object as its delegate.
    /// Safe to call more than once, from any queue.
    private func teardown() {
        lock.lock()
        let timer = _connectTimer
        _connectTimer = nil
        let s = session
        session = nil
        task = nil
        lock.unlock()

        timer?.cancel()
        s?.finishTasksAndInvalidate()
    }

    /**
     Number of leading bytes of `data` that form only complete UTF-8 sequences.

     Scans back at most four bytes for the last lead byte; if the sequence it
     starts is short of the length its prefix declares, that sequence is held
     back for the next delivery.
     */
    private static func completeByteCount(_ data: Data) -> Int {
        guard !data.isEmpty else { return 0 }
        let count = data.count
        var offset = 0

        while offset < 4 && offset < count {
            let b = data[data.startIndex + count - 1 - offset]
            let isContinuation = (b & 0xC0) == 0x80
            if !isContinuation {
                let need: Int
                if b & 0x80 == 0        { need = 1 }
                else if b & 0xE0 == 0xC0 { need = 2 }
                else if b & 0xF0 == 0xE0 { need = 3 }
                else if b & 0xF8 == 0xF0 { need = 4 }
                else                     { need = 1 }  // invalid lead: let the decoder replace it
                let have = offset + 1
                return have >= need ? count : count - have
            }
            offset += 1
        }
        // Four or more continuation bytes with no lead in sight: malformed, so
        // hand it all to the decoder and let it substitute.
        return count
    }

    /// Decode as much of `pending` as forms whole characters and hand it to JS.
    private func drain(reportingBytes byteLength: Int) {
        let safe = Self.completeByteCount(pending)
        guard safe > 0 else { return }  // whole delivery was a partial sequence

        let decodable = pending.prefix(safe)
        pending.removeFirst(safe)

        // `String(decoding:as:)` substitutes U+FFFD instead of returning nil, so
        // a malformed byte costs one character rather than the whole chunk.
        let text = String(decoding: decodable, as: UTF8.self)
        if !text.isEmpty {
            onChunk?(text, byteLength)
        }
    }
}

// MARK: - URLSessionDataDelegate

extension SseConnectionSwift: URLSessionDataDelegate {

    public func urlSession(
        _ session: URLSession,
        dataTask: URLSessionDataTask,
        didReceive response: URLResponse,
        completionHandler: @escaping (URLSession.ResponseDisposition) -> Void
    ) {
        guard !isCancelled else { completionHandler(.cancel); return }
        guard let http = response as? HTTPURLResponse else {
            completionHandler(.cancel)
            onError?("Non-HTTP response", -1, "NETWORK_ERROR", true)
            return
        }

        let status = http.statusCode
        guard (200 ..< 300).contains(status) else {
            completionHandler(.cancel)
            onError?("HTTP \(status)", status, "HTTP_ERROR", true)
            return
        }

        markOpened()

        var respHeaders = [String: String]()
        http.allHeaderFields.forEach { respHeaders["\($0.key)"] = "\($0.value)" }
        onOpen?(status, respHeaders)
        completionHandler(.allow)
    }

    public func urlSession(
        _ session: URLSession,
        dataTask: URLSessionDataTask,
        didReceive data: Data
    ) {
        guard !isCancelled else { return }
        pending.append(data)
        drain(reportingBytes: data.count)
    }

    public func urlSession(
        _ session: URLSession,
        task: URLSessionTask,
        didCompleteWithError error: Error?
    ) {
        guard !isCancelled else { teardown(); return }

        // Flush any trailing partial sequence so the server's last bytes are not
        // swallowed; the decoder substitutes for whatever is truncated.
        if !pending.isEmpty {
            let text = String(decoding: pending, as: UTF8.self)
            pending.removeAll(keepingCapacity: false)
            if !text.isEmpty { onChunk?(text, 0) }
        }

        defer { teardown() }

        guard let err = error as NSError? else { onClose?(); return }
        if err.code == NSURLErrorCancelled { return }
        let code = err.code == NSURLErrorTimedOut ? "TIMEOUT_ERROR" : "NETWORK_ERROR"
        onError?(err.localizedDescription, -1, code, false)
    }

    /**
     Redirects are followed, but credentials are not carried to a different
     origin: a server-chosen `Location` must not be able to harvest the
     Authorization header the caller set for the original host.
     */
    public func urlSession(
        _ session: URLSession,
        task: URLSessionTask,
        willPerformHTTPRedirection response: HTTPURLResponse,
        newRequest request: URLRequest,
        completionHandler: @escaping (URLRequest?) -> Void
    ) {
        guard !isCancelled else { completionHandler(nil); return }

        let sameOrigin =
            request.url?.host == url.host &&
            request.url?.scheme == url.scheme &&
            request.url?.port == url.port

        if sameOrigin {
            completionHandler(request)
            return
        }

        var stripped = request
        for header in ["Authorization", "Cookie", "Proxy-Authorization"] {
            stripped.setValue(nil, forHTTPHeaderField: header)
        }
        completionHandler(stripped)
    }
}
